src/Engine/GameAction.java

```diff
@@ -828,6 +831,100 @@ public UnitActionType getUnitActionType()
     }
   } // ~ResupplyAction
 
+  // ===========  RepairUnitAction  ===============================
+  public static class RepairUnitAction implements GameAction
+  {
+    private Path movePath;
+    private XYCoord startCoord;
+    private XYCoord moveCoord;
+    private XYCoord repairCoord;
+
+    public RepairUnitAction(XYCoord startLoc, Path path, int targetX, int targetY)
+    {
+      this(startLoc, path, new XYCoord(targetX, targetY));
+    }
+
+    public RepairUnitAction(XYCoord startLoc, Path path, XYCoord repairLoc)
+    {
+      movePath = path;
+      startCoord = startLoc;
+      repairCoord = repairLoc;
+      if( null != path && (path.getEnd() != null) )
+      {
+        moveCoord = new XYCoord(movePath.getEnd().x, movePath.getEnd().y);
+      }
+    }
+
+    @Override
+    public GameEventQueue getEvents(MapMaster gameMap)
+    {
+      // Repair actions consist of
+      //   MOVE
+      //   HEAL
+      //   REPAIR
+      GameEventQueue repairEvents = new GameEventQueue();
+
+      boolean isValid = true;
+      Unit benefactor = null;
+      Unit beneficiary = null;
+
+      if( (null != gameMap) && (null != startCoord) && (null != repairCoord) &&
+          gameMap.isLocationValid(startCoord) && gameMap.isLocationValid(repairCoord) )
+      {
+        benefactor = gameMap.getLocation(startCoord).getResident();
+        beneficiary = gameMap.getLocation(repairCoord).getResident();
+        isValid &= benefactor != null && !benefactor.isTurnOver;
+        isValid &= (null != beneficiary) && !benefactor.CO.isEnemy(beneficiary.CO);
+        isValid &= (null != gameMap) && (gameMap.isLocationValid(repairCoord));
```

Comment:
Shouldn't you make sure the location is valid before getting the resident on line 875? The `null == gameMap` check is redundant, and the other should probably also go in the `if` condition.

Reply:
The entire line is redundant, actually. The other is already in the `if`.
At least this copy-paste just wastes cycles, haha